File: candy.cpp
// https://leetcode.com/problems/candy/

#include <algorithm>
#include <vector>

class Solution {
public:
  int candy(std::vector<int> &ratings) {
    int n = ratings.size();
    if (n == 1)
      return 1; // Only one child gets one candy

    int totalCandies = 0;
    int leftToRightCandies = 1;
    std::vector<int> leftToRight(
        n, 1); // Store left-to-right candies in a single array

    // First pass: left to right
    for (int i = 1; i < n; ++i) {
      if (ratings[i] > ratings[i - 1]) {
        leftToRight[i] = leftToRight[i - 1] + 1;
      }
    }

    // Second pass: right to left and accumulate total candies
    int rightToLeftCandies = 1;
    totalCandies += leftToRight[n - 1]; // Start with the last child's candies

    for (int i = n - 2; i >= 0; --i) {
      if (ratings[i] > ratings[i + 1]) {
        rightToLeftCandies++;
      } else {
        rightToLeftCandies = 1;
      }
      totalCandies += std::max(leftToRight[i],
                               rightToLeftCandies); // Maximum of both passes
    }

    return totalCandies;
  }
};

File: candy.py
# https://leetcode.com/problems/candy

ratings = [0, 1, 2, 3, 2, 1]


def candy(ratings):
    n = len(ratings)
    candies = [1] * n  # Each child gets at least one candy initially

    # First pass: left to right
    for i in range(1, n):
        if ratings[i] > ratings[i - 1]:
            candies[i] = candies[i - 1] + 1

    # Second pass: right to left
    for i in range(n - 2, -1, -1):
        if ratings[i] > ratings[i + 1]:
            candies[i] = max(candies[i], candies[i + 1] + 1)

    return sum(candies)


print(candy(ratings))
